File: src/parse/detect.rs
/// Detected subscription format.
pub enum SubFormat {
    ClashYaml,
    Base64Uri,
    PlainUri,
}

/// Detect subscription body format.
pub fn detect_format(body: &str) -> SubFormat {
    let trimmed = body.trim();

    // Clash YAML: starts with common Clash config keys or contains "proxies:"
    if trimmed.starts_with("proxies:")
        || trimmed.starts_with("port:")
        || trimmed.starts_with("mixed-port:")
        || trimmed.contains("\nproxies:")
    {
        return SubFormat::ClashYaml;
    }

    // Plain URI lines: starts with a known proxy scheme
    if trimmed.starts_with("ss://")
        || trimmed.starts_with("ssr://")
        || trimmed.starts_with("vmess://")
        || trimmed.starts_with("vless://")
        || trimmed.starts_with("trojan://")
        || trimmed.starts_with("hysteria2://")
        || trimmed.starts_with("hy2://")
        || trimmed.starts_with("tuic://")
    {
        return SubFormat::PlainUri;
    }

    // Otherwise assume base64-encoded URI list
    SubFormat::Base64Uri
}

File: zig/src/parse/detect.zig
const std = @import("std");

pub const SubFormat = enum {
    clash_yaml,
    base64_uri,
    plain_uri,
};

pub fn detectFormat(body: []const u8) SubFormat {
    const trimmed = std.mem.trim(u8, body, " \t\r\n");

    if (std.mem.startsWith(u8, trimmed, "proxies:") or
        std.mem.startsWith(u8, trimmed, "port:") or
        std.mem.startsWith(u8, trimmed, "mixed-port:") or
        std.mem.indexOf(u8, trimmed, "\nproxies:") != null)
    {
        return .clash_yaml;
    }

    if (std.mem.startsWith(u8, trimmed, "ss://") or
        std.mem.startsWith(u8, trimmed, "ssr://") or
        std.mem.startsWith(u8, trimmed, "vmess://") or
        std.mem.startsWith(u8, trimmed, "vless://") or
        std.mem.startsWith(u8, trimmed, "trojan://") or
        std.mem.startsWith(u8, trimmed, "hysteria2://") or
        std.mem.startsWith(u8, trimmed, "hy2://") or
        std.mem.startsWith(u8, trimmed, "tuic://") or
        std.mem.startsWith(u8, trimmed, "anytls://"))
    {
        return .plain_uri;
    }

    return .base64_uri;
}
